### PR TITLE
Add cleanup transformation sorting inputs of commutative operations

### LINKED ISSUE #84: Add cleanup transformation sorting inputs of commutative operations
## Details
Most graph transformations (here as well as in FINN) assume a particular order of node inputs, even if the operation is actually commutative, e.g. the add operation. There seems to be a clear distinction between "dynamic" (i.e., produced upstream) and initializer inputs. As far as I can tell, the assumption is always that dynamic inputs are listed first, followed by the initializers. Concretely, `node.input[0]` is used to refer to the (single) dynamic input to node. However, this does not always hold true and I have encountered multiple occasions where valid transformations did not apply or were applied incorrectly due to violations of this assumption: https://github.com/Xilinx/finn/issues/878

### New behavior
I propose to add a cleanup transformation to sort the node input list of such commutative operations to have the initializer inputs last. For all other types of transformations, input order already has special meaning and seems to be handled correctly.

### Motivation
Ideally, transformations of commutative operations should not care for the input order. But as order is assumed in a lot of places right now, it seems to be easier to introduce a cleanup transformation to ensure the assumptions hold true. 

### Parts of QONNX being affected
A new cleanup transformation will be introduced and will be added to the default cleanup transformations of  the `ModelWrapper`.

### PR DESCRIPTION
**Pull request by iksnagreb**
Adds the cleanup transformation requested in #84

Closes #84

### CODE REVIEW

**Comment by maltanar** [comment]
Hi Christoph,

Thanks for the PR. I agree that this would be a useful cleanup transformation, but one remark after a brief review: since this gets called as part of the `ModelWrapper.cleanup()` which essentially gets called after every single transformation, and since the ONNX export behavior we observe from e.g. PyTorch has changed over the course of versions, we should try to add some extra checks for future-proofing here. Could you provide a unit test that exercises this transformation? Might be easiest to take one of the existing ONNX testcases and artificially manipulate the input list as part of the testcase such that the initialized input appears first.

File: src/qonnx/transformation/general.py [hunk 1]
```diff
@@ -359,3 +362,57 @@ def apply(self, model):
...
+            # Check whether this node is among the supported
+            if node.op_type in self.SUPPORTED_COMMUTATIVE_OPS:
+                # Group node inputs by category
+                dynamics, initializers = group_inputs_by_category(node, model)
+                # Flatten the grouped input list
+                inputs = [*dynamics, *initializers]
+                # Length of sorted and original input list must match
+                assert len(inputs) == len(node.input)
+                # Reassigned inputs from sorted categories
+                #   Note: ONNX does not allow direct assignment to node.input
```
Comment (maltanar) [inline src/qonnx/transformation/general.py:408]: minor remark for future info: although I haven't tried it, I believe it should be possible to assign directly to protobuf containers such as `node.input` by using the `node.input[:] = ...` syntax, similar to this:

https://github.com/fastmachinelearning/qonnx/blob/main/src/qonnx/custom_op/base.py#L127

Comment (iksnagreb) [reply src/qonnx/transformation/general.py]: Ok, thank you for pointing this out, I tried this somewhere else and this indeed works. Removed the comment here, but still kept the loop for assigning, as this simultaneously tests whether the graph has been modified (could probably be written less verbose, but I see now reason to change it now).

**Comment by iksnagreb** [comment]
> Could you provide a unit test that exercises this transformation?

Has been added now, testing this transformation on the `Sum` operator and some input/initializer combinations.

**Pull request merged**